cdap-master/src/main/java/co/cask/cdap/data/tools/ExistingEntitySystemMetadataWriter.java

```diff
@@ -0,0 +1,155 @@
+ * the License.
+ */
+
+package co.cask.cdap.data.tools;
+
+import co.cask.cdap.api.ProgramSpecification;
+import co.cask.cdap.api.app.ApplicationSpecification;
+import co.cask.cdap.api.data.stream.StreamSpecification;
+import co.cask.cdap.api.dataset.DatasetProperties;
+import co.cask.cdap.app.store.Store;
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.data.dataset.SystemDatasetInstantiatorFactory;
+import co.cask.cdap.data.view.ViewAdmin;
+import co.cask.cdap.data2.dataset2.DatasetFramework;
+import co.cask.cdap.data2.dataset2.DatasetManagementException;
+import co.cask.cdap.data2.metadata.store.MetadataStore;
+import co.cask.cdap.data2.metadata.system.AbstractSystemMetadataWriter;
+import co.cask.cdap.data2.metadata.system.AppSystemMetadataWriter;
+import co.cask.cdap.data2.metadata.system.ArtifactSystemMetadataWriter;
+import co.cask.cdap.data2.metadata.system.DatasetSystemMetadataWriter;
+import co.cask.cdap.data2.metadata.system.ProgramSystemMetadataWriter;
+import co.cask.cdap.data2.metadata.system.StreamSystemMetadataWriter;
+import co.cask.cdap.data2.metadata.system.ViewSystemMetadataWriter;
+import co.cask.cdap.data2.transaction.stream.StreamAdmin;
+import co.cask.cdap.internal.app.runtime.artifact.ArtifactDetail;
+import co.cask.cdap.internal.app.runtime.artifact.ArtifactStore;
+import co.cask.cdap.proto.DatasetSpecificationSummary;
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.proto.NamespaceMeta;
+import co.cask.cdap.proto.ProgramType;
+import co.cask.cdap.proto.artifact.ArtifactInfo;
+import co.cask.cdap.store.NamespaceStore;
+import com.google.inject.Inject;
+import org.apache.twill.filesystem.LocationFactory;
+
+import java.io.IOException;
+import java.util.Collection;
+
+/**
+ * Adds system metadata for existing entities prior to 3.3.
+ * TODO CDAP-4696: Remove in 3.4
+ */
+public class ExistingEntitySystemMetadataWriter {
+  private final MetadataStore metadataStore;
+  private final NamespaceStore nsStore;
+  private final Store store;
+  private final StreamAdmin streamAdmin;
+  private final ViewAdmin viewAdmin;
+  private final ArtifactStore artifactStore;
+  private final LocationFactory locationFactory;
+  private final CConfiguration cConf;
+
+  @Inject
+  ExistingEntitySystemMetadataWriter(MetadataStore metadataStore, NamespaceStore nsStore, Store store,
+                                     ArtifactStore artifactStore, StreamAdmin streamAdmin, ViewAdmin viewAdmin,
+                                     LocationFactory locationFactory, CConfiguration cConf) {
+    this.metadataStore = metadataStore;
+    this.nsStore = nsStore;
+    this.store = store;
+    this.streamAdmin = streamAdmin;
+    this.viewAdmin = viewAdmin;
+    this.artifactStore = artifactStore;
+    this.locationFactory = locationFactory;
+    this.cConf = cConf;
+  }
+
+  public void write(DatasetFramework dsFramework) throws Exception {
+    for (NamespaceMeta namespaceMeta : nsStore.list()) {
```

Comment:
Do you think it is worth it to wipe out all the system metadata first before adding new stuff? That way, we can re-use this tool even after 3.3.x, if/when we change the system metadata that we annotate for all entities.